extensions/security/runtime/src/main/java/io/quarkus/security/runtime/interceptor/SecurityHandler.java

```diff
@@ -20,8 +27,42 @@ public Object handle(InvocationContext ic) throws Exception {
         if (alreadyHandled(ic)) {
             return ic.proceed();
         }
-        constrainer.check(ic.getMethod(), ic.getParameters());
-        return ic.proceed();
+        Class<?> returnType = ic.getMethod().getReturnType();
+        if (Uni.class.isAssignableFrom(returnType)) {
+            return constrainer.nonBlockingCheck(ic.getMethod(), ic.getParameters())
+                    .onItem().transformToUni(new Function<Object, Uni<?>>() {
+                        @Override
```

Comment:
Minor one, may be this function can be an internal class parameterized by its return type, and statically created for these 3 types, in case creating a new instance every time might be of some concern

Reply:
If you'd like you can resolve it, I don't really mind how it is handled, is probably very optimized once compiled with the functions

Reply:
We have a ton of functions in the OIDC code the only difference is that they are all unique in their structure, but here they are the same hence I suggested it :-)

Reply:
not sure it changes a lot but I pulled them out